gms/gossiper.hh

```diff
@@ -76,6 +76,19 @@ class i_endpoint_state_change_subscriber;
 
 class feature_service;
 
+struct gossip_query_token_status_response {
+    // Number of nodes this node knows about
+    uint64_t nr_known_nodes;
+    // The hash of the tokens for nodes in NORMAL status
```

Comment:
> Should we also take into account nodes in the SHUTDOWN status?

The SHTUDOWN node will have the tokens calculated in the normal_token_hash, because it is part of the token ring. 
> 
> Why do we need to split the hashes of tokens based on statuses and not just hash all the tokens, regardless of the status?

We split the token based on: tokens in the token ring and new tokens will be added to the token ring.
node1: t1, t2 (not in the token ring yet)
node1: t1, t2 (in the token ring)

If we do not split, this will give the same hash while the cluster status is very different.

Reply:
> > Why do we need to split the hashes of tokens based on statuses and not just hash all the tokens, regardless of the status?
> 
> We split the token based on: tokens in the token ring and new tokens will be added to the token ring.
> node1: t1, t2 (not in the token ring yet)
> node1: t1, t2 (in the token ring)
> 
> If we do not split, this will give the same hash while the cluster status is very different.

What does "not in the token ring yet" mean?

Isn't ring membership determined by node's STATUS, in which case the hash of the statuses takes care of detecting the difference?

Reply:
The discussion was moved to standalone comments below. Closing it here.